### ISSUE TITLE
Central Bank Digital Currency (CBDC): towards proven technology, ready for large-scale deployment

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#6306]
The goal of this master thesis research is to take the necessary steps to mature Central Bank Digital Currency (CBDC).
The level of maturity of CBDC targeted within this thesis is "proven technology, ready for large-scale deployment". 

Stress testing and integration with the upcoming digital European identity is expected to be a significant part of this thesis work. All software will be open source and all gained technical expertise will be shared publicly. The permissionless ledger technology used will be Trustchain, this is [formalised an IETF Internet Standard draft since 2018](https://tools.ietf.org/id/draft-pouwelse-trustchain-01.html). The solution is required to be ledger agnostic.

This CBDC technology is required to be fully decentralised. Inefficient and wasteful "mining" is explicitly not allowed. No critical dependency on existing banking infrastructure may exist. No central points of failure or performance bottleneck can be present in the system architecture. Due to the critical nature of financial infrastructure, an "offline mode" must be supported. Transfer of money using QR-codes or some other mechanism must be supported together with fraud measures.

Delft University of Technology has already conducted small trails with digital Euros, supporting offline transactions. The [work by Wessel Blokzijl](https://github.com/Tribler/tribler/issues/4629) resulted in a fully functional prototype, featuring real Euros, real retail testing, and integration with the IBAN banking system using a "Tikkie"-based gateway. [Read the full master thesis here](https://repository.tudelft.nl/islandora/object/uuid%3A132faae8-6883-454f-a8ce-94735340dce9). Delft University of Technology is also a government partner for digital identity at passport-level. We have an operational open source prototype for digital identity, integrated with the European Commission EBSI infrastructure. By using EU EBSI this thesis aims for seamless integration of identity and money. See: https://github.com/Tribler/tribler/issues/6023 This existing self-sovereign identity work of Delft will be re-used.

Stress testing is also a key part of demonstrating the maturity of this technology based on open source and open standards. The requirement is horizontal scalability, feasible due to the inherent parallelism of the used Trustchain ledger. The ambitious target for experimental demonstration is 1 million transactions per second. This will provide the irrefutable proof required that a digital Euro can underpin the entire digital economy of Europe. The check-pointing mechanism used by Wessel might need replacing or improvement. Finally, a small field trail (4 people) would demonstrate the end-to-end feasibility and maturity level of this system. In the ideal case CBDC would utilise an IBAN account owned by the central bank to conduct several 1 Euro transactions in the wild.  Possible joint experiments of other governments will be explored with actual money, but very low amounts. For instance, exchange of 1 actual Euro into an equivalent Singapore Dollars, e-Kroner (Sweden) or collaboration with the German Bundesbank to conduct end-to-end system testing.

**Comment by KoningR** [comment]
Progress Report 
- Build an run Trustchain app
- Read Blokzijl thesis
- Read ValueTransfer thread
- Interface crashed when transferring money; made issue report on Git
- Read interview with Fabio Panetta (Der Spiegel)
- Read EU report on digital euro (about 60%)
    - Largely answers question "why a digital euro"

**Comment by synctext** [comment]
2020 work by Jetse: https://repository.tudelft.nl/islandora/object/uuid%3Ad3d56dd8-60ee-47f7-b23a-cdc6c2650e14

**Comment by KoningR** [comment]
Sprint log (last week)
- Read Consensus-less Security (Brouwer).
- Inspect and (partially) understand Trustchain/Eurotoken/Superapp code.
- Attempt to set up own gateway using Blokzijl's repo.
- Failed; instructions were mostly non-existent. Current problem: I do not have a Tikkie API key.

**Comment by synctext** [comment]
Comments. [This is an important CBDC report from China](http://www.pbc.gov.cn/en/3688110/3688172/4157443/4293696/2021071614584691871.pdf)
```
The People’s Bank of China (PBOC) attaches great importance to the research and
development of digital fiat currency. In 2014, it set up a task force to study digital fiat
currency, and its scope of research covered the issuance framework, key technologies,
issuance and circulation environment, and international experience.
```
ECB has started the task force and announced their "digital Euro" exploration. Still, Europe is 7 years behind.

**Comment by synctext** [comment]
First _live_ digital Euro transaction?! Worldwide state-of-the-art, parties are mentioning cross-border CBDC; never actually achieved yet! 
- See Hong Kong Monetary Authority ("proof-of-concept" level only): "we are delighted to join forces with the Bank of Thailand in conducting a joint research project named “Project Inthanon-LionRock” to study the application of CBDC to [cross-border payments, with a view of facilitating HKDTHB PvP amongst banks in Hong Kong and Thailand.](https://www.hkma.gov.hk/media/eng/doc/key-functions/financial-infrastructure/Report_on_Project_Inthanon-LionRock.pdf)
- JP Morgan made the first "stable coin" claims years [ago](https://www.coindesk.com/markets/2019/02/14/jpmorgan-has-its-own-crypto-and-its-starting-real-world-trials/), always vaporware. Now part of Consensys CBDC Ethereum. Still proof-of-concept level: "will involve the development of a POC for the issuance of a tokenized form of CBDC", https://consensys.net/blog/enterprise-blockchain/cbdc-pilots-on-ethereum/
- France: Societe Generale, live transaction “explores a more efficient process for bond issuances.” issued a €100 million covered bond as a security token on the Ethereum blockchain https://tokenist.com/societe-generale-issues-e100-million-covered-bond-as-a-security-token-on-ethereum/

**Comment by synctext** [comment]
ideas: permissionless gateways. Everybody can undersign a transaction, remove single points of failure.

**Comment by KoningR** [comment]
https://drive.google.com/file/d/1hGoFH6gjEh4OQbxpWdlZ5NxYYtS2m0KA/view?usp=sharing

**Comment by KoningR** [comment]
- Prepared current Eurotoken code for a demo at TU Delft using a remote server on a different network.
- Unexpectedly stumbled upon a bug that prevents Trustchain from working on certain mobile data providers; this was known already and (verified) not due to Eurotoken.
- Discussed potential directions of Eurotoken at TU Delft.

**Comment by synctext** [comment]
- **2050 vision**: disaster-proof financial infrastructure. Very few failure scenarios, reliable, robust and scalable infrastructure. Desired property: security is improved with increased usage. Provide underpinning for the robot economy.\
- towards 1 million transactions per second. Great quantitative material. Isolated Kotlin code. 
- Counter-party risk assessment function: spend _2 weeks_ on this maximum! Do a simple or even trivial implementation of this. Use history to predict future reliability. Leave for future work.

**Comment by KoningR** [comment]
### Sprint log
- Built kotlin-ipv8 locally (which was an unnecessarily complicated endeavour due to a lack of documentation).
- Made a [pull request](https://github.com/Tribler/kotlin-ipv8/pull/55) to update the readme with the information I would have liked to know before starting the building process.
- Found a bug that prevents Trustchain from working at all when executed from command line using kotlin-ipv8. Made an [issue](https://github.com/Tribler/kotlin-ipv8/issues/56).
- Fixed this issue with this [PR](https://github.com/Tribler/kotlin-ipv8/pull/57).
- Implemented a naive communication protocol to send Eurotoken using Trustchain, assuming only honest nodes.

**Comment by synctext** [comment]
- Solid progress with the infrastructure! :medal_sports: 
- further narrow down the scientific direction (offline; 1M tx/s) and sprint target.
- Serial number of bank coins as central to the design? Single signature ownership or dual signatures? <IMG src="https://user-images.githubusercontent.com/325224/142997663-5b34f426-5d56-45ba-bc56-ba12325e2d88.png" width=200>

**Comment by KoningR** [comment]
### Sprint log
- Modified Eurotoken database schema to support grouped transactions and query transaction history per coin or peer.
- Modified the command line application to accept text files containing instructions.
- Attempted unsuccessfully to debug why a Eurotoken peer sometimes receives the same genesis block twice.
- Attempted to measure throughput but failed due to NEXT_HASH_MISMATCH. Currently investigating which mechanisms should be used to circumvent this error.

**Comment by synctext** [comment]
Progress: 
- cmdline is functional!
- https://github.com/KoningR/eurotoken/blob/main/README.md no build instructions

**Comment by KoningR** [comment]
## Sprint log
- Read Matt's thesis.
- Read De Vos's Xchange paper.
- Once again attempted to fix SQL duplicate block error without success.
- Called with Rowdy and discussed RUDP and delays.
- Implemented a simple delay mechanism to gauge the throughput of Trustchain.
- Measured Trustchain's throughput and plotted this data. The measured performance is, in many ways, not representative. Here's why: 
     - Measurements were done using 2 peers, connected over LAN. 
     - There is no verification mechanism in place, so 'throughput' effectively means the roundtrip speed of UDP between the clients. 
     - There is no retrieval mechanism for UDP in place, so 'lost' eurotokens were never retrieved again.
     - To reduce the number of lost messages, the sending client was restricted to a waiting time that was determined using a rudimentary binary search.
     - In short, there is no fixed eurotoken protocol yet, and as such, a comparison of performance does not make much sense.
     - <img src="https://user-images.githubusercontent.com/23524378/148095075-4663ad0e-1dcd-4319-a653-4e497f1c8593.png" alt="performance plot" width="300" height="300">
- Made a eurotoken poster to display on campus.
- Added build instructions to [Eurotoken](https://github.com/KoningR/eurotoken).
- Made a [PR](https://github.com/Tribler/kotlin-ipv8/pull/58) to improve the build instructions of IPv8.

**Comment by synctext** [comment]
- Poster: [first trail March 2021](https://github.com/Tribler/tribler/issues/4629#issuecomment-789934358) of Delft offline Euro platform. Based on Trustchain and self-sovereign identity

**Comment by synctext** [comment]
Related CBDC work with instant payments: https://www.bis.org/publ/othp39.htm

**Comment by KoningR** [comment]
## Spring log
- Work on merging the dispatcher fix with the new EVA protocol and figure out why this breaks the trustchain superapp.
     - Contacted Joost for this.
- Visit DNB to meet my thesis supervisors and co.
     - Set up a work laptop and read a lot of mails.
     - Had small introduction meetings via Teams.
- Re-read Wessel's paper to figure out what exactly did and didn't work.

**Comment by synctext** [comment]
- Contract 15jan-15july at central bank.
- define scope of thesis

| Month | activity |outcome |
|---|---|---|
|Jan| | |
|Feb| | |
|March| | |
|April| | |
|May| | |
|June| | |
|July| | |

**Comment by KoningR** [comment]
## Sprint log
- Wrote a draft version of the Eurotoken protocol to identify theoretical limitations. 
    - Called with Rowdy to discuss this draft.
- Modified Eurotoken implementation to use EVA instead of Trustchain. 
- Attempted to fix a dispatcher problem, unsuccessfully.
- Measured performance of EVA instead of Trustchain with manual delays. The good: no more lost packets. The bad: EVA appears to perform slower than a manual sending delay.
- Called with various people at DNB and attended a digital session about policies for DLT.

**Comment by synctext** [comment]
- Progress in thinking, requirements, and architecture
  - Offline usage is critical requirement
  - Online mode 
- zero-trust architecture, implemented in USA. Discussed in [presidential memo](https://www.bastionzero.com/blog/i-read-the-federal-governments-zero-trust-memo-so-you-dont-have-to). ```a Federal zero trust architecture (ZTA) strategy, requiring agencies to meet specific cybersecurity standards and objectives by the end of Fiscal Year (FY) 2024```
- Focus and scope (repeating and combining two issues above):
  1. rudimentary performance mode (its impossible to go faster; linear scalability?)
  2. online mode (quantify the performance cost of safety features?)
  3. offline mode (1 risk of effort in risk assessment module to inform users)
- correctness: 25 tx per second for 2nd prototype.

**Comment by synctext** [comment]
- China CBDC usage. They are [distributing HK$5000 of money that expires](https://www.chinadailyhk.com/article/231112). Economic booster instrument, [helicopter money](https://en.wikipedia.org/wiki/Helicopter_money).
- USA: "This collaboration between MIT and our technologists has created a scalable CBDC research model that allows us to [learn more about these technologies and the choices that should be considered when designing a CBDC](https://www.bostonfed.org/news-and-events/press-releases/2022/frbb-and-mit-open-cbdc-phase-one.aspx)."

**Comment by KoningR** [comment]
## Sprint log
- Read Joost's thesis.
- Re-read this entire ticket.
- Read about colored coins.
- Made a separate repo for the plots.
- Measured the EVA protocol with various batch sizes and found a HUGE performance increase.
    - The plots also seem to indicate that data transfer speed is the bottleneck; persistence barely makes a dent.
    - <img src="https://user-images.githubusercontent.com/23524378/154339936-23ebd224-9413-427b-bbb9-1f790aab482b.png" alt="eva_total_10000_batch_500" width="200" height="200">
    - <img src="https://user-images.githubusercontent.com/23524378/154340161-dfc1e60b-c3c7-47f7-9ec0-799ffcc34d39.png" alt="eva_total_10000_batch_10000" width="200" height="200"> 
- Created a StressCommunity and analysed ipv8's sending routines.
    - The setup: a laptop sent UDP packets using ipv8 over Wi-Fi to a desktop connected via LAN.
    - Internal discussion about token sizes and throughput led to changing the units of the graphs to UDP packets (1472 usable unencrypted bytes) instead of 'tokens'.
 
   - When sending UDP without ACKs:
       - Recreating a fixed, simple payload gives a throughput of ~6.5mbps.
       - Reusing the same payload is much faster but causes many lost messages (almost 90%).
        - When sent in very small batches, the throughput is more than twice as high (~15mbps).
        - This suggests that on weaker CPUs the impact of creating packets is non-negligible.
        - It might be that reusing the same payload accelerates transfer so much that somewhere a buffer fills up and many packets are dropped.
        - Drops become more prevalent with larger batches, suggesting that indeed a buffer overflows.
    - Because reusing the same payload is not realistic, I attempted to figure out what caused new payloads to be so much slower. I identified (potential) slowing causes in ipv8, modified them, and measured the results.
        - I modified IPV4Address to create an INetAddress once upon creating instead of once per packet send.
        - I removed the multithreading for each individual packet because Java's UDP send methods are synchronized blocks.
        - I modified ipv8's serialize() method to be more memory effective.
        - <img src="https://user-images.githubusercontent.com/23524378/154340436-caad72a7-dec0-4467-847e-16125ebc5bf9.png" alt="udp_to_desktop" width="300" height="300">
- I then measured the performance of EVA with default settings, which sometimes gave okay results (~1.5mbps) but regularly faced enormous outliers (~100kbps).
 
- Conclusion 1: there needs to be a discussion regarding the size of a token, because it influences throughput drastically.
- Conclusion 2: ipv8's UDP and signing functionality appears fast enough to scale to Eurotoken's demands. EVA's does not seem to. However, it must be mentioned that no extensive parameter search was done in this domain.

**Comment by synctext** [comment]
- Solid progress since prior sprint
- Try to get Kotlin Native operational, limited to 1 sprint attempt only.
- Key reason is to eliminate magic
  - > It might be that reusing the same payload accelerates transfer so much that somewhere a buffer fills up and many packets are dropped.
  - At master thesis level we require exact answers. Socket buffers should be profiled. IPC as baseline. Go beyond 100Mbps.

**Comment by synctext** [comment]
- advise: try to think of a time-boxed approach, using 18 days since last meeting for IPC could be improved. 
- Determine cost-of-crypto. Profiling of full-speed verify and sign libraries.

**Comment by KoningR** [comment]
https://github.com/Tribler/kotlin-ipv8/blob/dc25de566120fabc5f2b1945e1e2198caecd5f5c/ipv8/build.gradle

**Comment by KoningR** [comment]
For a (slightly more) documented Wessel gateway refer to [my fork](https://github.com/KoningR/stablecoin-exchange). Do note that I am not maintaining this fork or the corresponding upstream repository.

**Comment by KoningR** [comment]
## Sprint Log
- Measured performance of IPC, UDP (native and JVM, reusing and recreating packets, local and LAN), and signing and verifying with LazySodium (1 batch and split batches).
    - Set up Ubuntu twice.
    - Wrote IPC and UDP code in C using Linux pipes.
    - Wrote UDP JVM and LazySodium code in Kotlin.
    - Performed all measurements 5 times.
    - Wrote plotting code in Python.
- <img src="https://user-images.githubusercontent.com/23524378/159715595-91bfe14a-cfc3-4a99-8b31-77fdd9ef4d6f.png" alt="upperbounds" width="200" height="200">
    
    - All measurements timed how long it took to process 100 megabytes of data.
    - The UDP measurements were performed by sending 200 megabytes of data, of which only 100 megabytes needed to be received.
        - No acknowledgement system was measured.
    - All measurements were done 5 times with 5 warmup rounds on a Lenovo Thinkpad.
    - LAN performance was measured using a 1 Gigabit connection. Measurements indicate that all LAN performance was limited around ~118 megabytes per second, close to the theoretical limit of 125.
    - Actions labeled 'reuse' indicate that the same UDP packet was used for all 100 megabytes of data. Actions labeled 'recreate' indicate that the UDP packet was recreated for each send.
    - Actions/protocols:
        - IPC: Inter-Process Communication in the form of named pipes implemented in C on Linux.
        - UDP(C-reuse): UDP implemented in C and reusing the same packet.
        - UDP (JVM-reuse): UDP implemented in Kotlin and reusing the same packet.
        - UDP (JVM-recreate): UDP implemented in Kotlin and recreating the packet each send.
        - UDP (JVM-recreate-LAN): Identical to UDP (JVM-recreate) but the sender and receiver process were different machines connected over LAN using a 1 Gigabit connection.
        - Sign: The signing performance of LazySodium (Kotlin) over 1 batch of 100 megabytes.
        - Verify: The verification performance of LazySodium (Kotlin) over 1 batch of 100 megabytes.
        - Sign (split): The signing performance of LazySodium (Kotlin) over batches of 1472 bytes, multithreaded (8 threads), for a total of 100 megabytes.
        - Verify (split): The verification performance of LazySodium (Kotlin) over batches of 1472 bytes, multithreaded (8 threads), for a total of 100 megabytes. On the measured system, this action was the slowest of the measured actions, with a throughput of ~90 megabytes per second.
    - The measurements performed are upper bounds for user performance, as it is highly unlikely that current smartphones have Gigabit peer-to-peer connections or processors fast enough to achieve comparable sign and verification performance.

**Comment by synctext** [comment]
_repeating_ Grand storyline: we take the existing ```WesselEuro```, dump all the code, redo stuff, and get it a step close to production usage. For that testing is essential.

**Comment by KoningR** [comment]
## Sprint log
- Drop threading merge because I don't use the superapp.
- Attempted to figure out why Kotlin-reuse is so much slower than C-reuse.
    - Recreated measurement code in Java to see if that was faster.
    - It was not.
- Before implementing performance mode, figured out how a valid minimal IPv8 packet should look like that carries tokens:
    - UDP header: 8
    - IPv8 prefix: 22
    - Recipient key: 74
    - Token: 1500 - 8 - 22 - 64 - 74 = 1332 bytes
    - Signature: 64
- Implemented measurements with Java NIO.
- While working on performance mode, took a deep dive into byte buffers.
    - Found a bug in my JVM implementation, will report soon.
- Went to the bank.
- Implemented performance mode; found too slow with ArrayList, LinkedHashSet, LinkedList.
    - With HashSet throughput of roughly 14.8 megabytes per second.
    - Currently working on optimizations.
- New measurement plot:
- <img src="https://user-images.githubusercontent.com/23524378/162638702-a2f4db05-84f5-4223-b84d-8bc9973fcb96.png" alt="upperbounds" width="300" height="300">

**Comment by synctext** [comment]
- What are you measuring? 100MByte test file on an Ubuntu laptop with 16GByte should be no problem. Usage of singleton?
- Profile sign and verify of digital cash
- Focus next sprint
  - Understand performance of offline mode.
  - Performance mode
  - Online mode: verifications per second of a gateway
    - Laptop and PC setup
    - Linux Laptop with 3 processes: send,receive, and validator
    - determine: tx/sec with check every micro-block and per 100 transactions

**Comment by synctext** [comment]
Related work by Timeguard.com
- ```during the test handled over 300,000 simultaneous payments a second and the money reached the payee in less than two seconds``` https://www.eestipank.ee/en/press/experiment-central-banks-digital-euro-illustrated-new-possibilities-blockchain-technology-26072021
-

**Comment by KoningR** [comment]
### Sprint log
- Implemented online mode compatible with IPv8!
  - Clients and verifier are operational.
- Bad news
    - 'Only' 2700 transactions per second per client.
        - Why?
          - Tokens need to contain public keys of sender and receiver (2 * 74 bytes) and a signature (64 bytes): 221 bytes per token (of which id and value take only 9). Assuming a download speed of 10 megaBYTES per second, no packet loss, no overhead, we are limited to 4525 transactions per second per client.
           - Packet loss is immense; need for acknowledgement system is dire.
             - Measurements largely measure packet drop, not actual performance.
             - EVA might not be suitable / too slow / build my own?
           - Cryptography per token is slow, but necessary.
              - Batching signatures is not an option because a signature should
refer to 1 token only.
               - Cryptography library is Java wrapped, does not support indexing and instead uses array copy. Presumably slow; not yet profiled.
            - Persistence mechanism will probably incur another performance hit.
- Optimism
  - Offline mode can potentially greatly increase the number of transactions by passing them on before having them verified.
  - Threading and coroutines should increase performance; not built in yet, first a persistence mechanism.

**Comment by synctext** [comment]
- solid performance results! All thesis-level material for master completion
- Graphs: UDP-C-reuse, UDP-Java-reuse, UDP-Kotlin-reuse, UDP-Kotlin-recreate, etc.
- Have a storyline for UDP packet loss
  - you dont mention things like rate control and congestion control. Blind UDP blasting :-) That is known for 30+ years to be insufficient.
  - Ongoing student work: https://github.com/Tribler/kotlin-ipv8/pull/63
  - Compare unconstrained UDP, EVA {with student fixes}, and spend maximum a few days on getting results from: https://github.com/benhenshaw/quick_udp/tree/master/examples
- Upcoming sprint: first draft
  - put everything in a single IEEE-format Latex paper
  - Chapters?? intro, problem description, design, experimental setup, performance analysis, related work, conclusions
  - 1-page intro of context. going cashless, CBDC, ```81% of Chinese consumers used Alipay```, https://hbr.org/2020/09/why-hasnt-apple-pay-replicated-alipays-success
  - 1 or 2 figures. "Performance analysis"

**Comment by KoningR** [comment]
## Sprint log
- Wrote introduction and problem description.
- Read various papers and articles
- Visited the bank and attended presentations about stable coins and crypto currencies.

[Introduction+ProblemDescription-24-05-22.pdf](https://github.com/Tribler/tribler/files/8763454/Introduction%2BProblemDescription-24-05-22.pdf)

**Comment by synctext** [comment]
- First writing of master thesis, 1 lap :checkered_flag: 
- Shorter, more powerful sentences
- Avoid merging 2020 report with a 2022 speech into a single sentence
- Example: Digitisation means the European Central Bank has less control on monetary policy matters.

**Comment by KoningR** [comment]
## Sprint log


 - Looked into adding QUIC as acknowledgement protocol.

    ○ The verdict: plenty of libraries but only 1 for JVM; not production-ready and poorly documented. Porting one of the other libraries seems like a major struggle (languages: C, Go, Rust).
 
- Measured the performance of the repo's serializePacket() vs. my own 'faster' version, turned out to not be faster but comparable. Conclusion: do not switch.
    
    ○ For 10000 simple serializations, with sign took ~567591600ns for the old version and ~571432400 for the new version. Without sign took ~17777300ns for the old version and ~18118100ns for the new.

        Sign:         567591600    571432400
        No sign:   17777300      18118100

- Reformatted the 'practical upperbounds graph' and 'cryptography' graph.
- Wrote a short description of the setup and described the 'upperbounds' graph and the 'cryptograhy' graph.
- Rewrote the introduction and problem description.
- Added references and included them in the text. Added more 'formal' sources such as statistics from DNB.

**Comment by KoningR** [comment]
[Thesis_Eurotoken_27-06-22.pdf](https://github.com/Tribler/tribler/files/8989368/Thesis_Eurotoken_27-06-22.pdf)

**Comment by synctext** [comment]
> ```An important difference is that Blokzijl’s system is balance based and this scenario is token-based.```

Problem descriptions are neutral. Not a single Blokzijl reference please. Build up the problem, then discuss timeguard and Blokzijl testing at 1 retail location (with photo) in _related work_. You can be very direct there: our work build fully upon this prior work. We aim to advance it further on the long road towards mass adoption.

**Comment by synctext** [comment]
https://nieuws.ing.nl/nl-NL/216305-ing-nxp-en-samsung-ontwikkelen-nieuwe-manier-van-onderling-betalen-met-mobiele-telefoon

ING is not giving any credits to TUDelft, but they are trying also P2P payments :2nd_place_medal: (deployed 2021 with live IBAN linkage). Requires special radio chip and special Samsung permissions. Good GUI and roadmap to mass deployment, positioning as alternative against ABN Tikkie. Digital Euro needs to match this market-based innovation!

**Comment by KoningR** [comment]
- Searched for a sensible statistic regarding counterparty risk; determined 'number of failed banks' an inadequate statistic; found little further usable material or sources. 
 - Debugged the EVA protocol to figure out why the throughput on 1 machine was less than 2 megabytes per second (16 mbps).
     - Attempted: reducing window ACK resend delay, removing encryption per packet, changing window size.
     - One potential reason appears to be that EVA encrypts data packets individually, which we have seen can have a large impact on performance.
     - One potential reason appears to be that EVA signs packets individually and verifying these signatures can have a large impact on performance.
     - After profiling and removing encryption and signatures, throughput was still low.
     - One of the potential reasons appears to be the long delay between resending a window acknowledgement. EVA is not a sliding window but a static window protocol; if a window ACK is lost, the system waits on average 1 second before resending. In the meantime, no packets are sent.
 - Measured the throughput of ipv8 because Bambacht speculated EVA throughput was limited due to ipv8. Found that raw ipv8 throughput in the test setup was around 20 megabytes per second (160 mbps). However, throughput dropped massively with encryption enabled, to 5 megabytes per second (40 mbps).
 - Learned that the EVA protocol asks for unreceived blocks in 1 packet that can get arbitrarily large. This makes increasing the window size impossible without changing the protocol code.
 - From the Python version of EVA:  "the protocol is slow by design".  <https://github.com/Tribler/tribler/blob/1fe7a9fc23c467f35fcffff71642abe2128da214/src/tribler/core/components/ipv8/eva/protocol.py>
     - Currently debating whether it is even worth the time to attempt to optimise a protocol that is deliberately slow. The result will be, however, that the Eurotoken is slow. 
 - Discussed with Martijn:
     - He is also looking into a faster acknowledgement system, but for Python.
     - One important factor is congestion control, which EVA does not have.
     - It seems unrealistic to improve upon or build from scratch and acknowledgement protocol with the remaining time and thesis objectives.
     - TL;DR I'm stuck with EVA and am to accept its throughput.
         - The good news: I do not have to worry about acknowledgements anymore.
         - The bad news: Throughput per-phone is low.
         - Optimism: throughput per-phone can be low as long as the entire system can support a large number of transactions.
 - Another issue: explain the choice for using tokens instead of balances.
     - Major drawback: increases required number of transactions due to fixed token value.
     - Advantages (discussed with Martijn): easy trackability. 
 - Tried to figure out if offline trackability can be done without using tokens. Turns out that using tokens is indeed a straightforward way of tracking offline transactions. 
 - Rewrote offline transaction protocol to account for replay attacks. 
 - Got Eurotoken to work with EVA. 
 - Implemented offline token transaction protocol. Double spending detecting will be finished within a day.

**Comment by synctext** [comment]
- https://www.fdic.gov/bank/historical/bank/ Reputable source and also clearly shows exceptional events with mass bank bankruptcies. Please do not spend too much time on this, perfectionism trap.

**Comment by KoningR** [comment]
## Sprint Log
- Described the token transaction protocol in a thesis-worthy manner.
- Made diagrams of token transaction protocol.
- Discovered a double spending attack possible in the protocol.
    - ○ The current protocol allows a malicious party to spend the exact same token to the exact same person at a later time. If the receiver of this doubly spent token is unaware of this and continues spending the token, later on when the token is redeemed, they are accused of double spending.
- Discussed with self and others about the possibilities of preventing such an attack.
    - ○ Each client maintains a database of all tokens (last hashes) ever received to prevent the attack. 
        - § Drawback: the database can grow near indefinitely and may never be lost.
    - ○ Before making a transaction, the sender first requests a nonce from the receiver which will be included in the token transaction such that double spending cannot be done.
        - § Drawback: The request adds a message roundtrip and the receiver must be online at the time of transacting.
    - ○ The sender includes a timestamp in their transaction. This timestamp must be somewhat synchronized with the clock of the receiving party. For a fixed duration, the receiver will not accept tokens from this specific sender if they contain the same timestamp. This option is currently preferred.
        - § Drawback: The receiver must be online at the time of transacting.
 - Wrote briefly about the choice to use tokens instead of balances.
 - Decided against using FDIC graph in thesis because the data refers to the US and does not distinguish between consumers and business, but most importantly because the ECB does not seem to consider substituting commercial banks as a goal of their CBDC. In fact, quite the opposite, they cannot allow their CBDC to replace commercial banks.
 - Would like to discuss: what do you suggest are proper sizes of token Id, Value, and Nonce? The other fields (public keys and signatures) were already decided in ipv8. Necessary to do before final experiment run. 
 - Came up a with a proper setup for a performance analysis:
     - ○ A brief section measuring the performance of the low-level functionality. Do not dedicate an entire figure (even though this work was already done :() but a table. The data says very little because almost all measurements are >125mbps, which is the highest attainable speed in a practical scenario.
     - ○ Perform new measurements for the cryptography section:
         - § Redo cryptography results and differentiate between verifier (on PC) and client (on phone, when implementation works in the app).    § Verifier:     □ Measure speed of cryptography verification in chunks of the same size as the token protocol uses.     □ Measure signing speed in chunks of the same size.    § Perhaps measure another native cryptography library on PC to determine throughput loss.   ○ Measure IPV8 send performance with and without encryption. Also measure packet loss.   ○ Next, measure EVA in different configurations.    § Why? Do I redo Bambacht's work then? What's the point?    § Simply copy the 'optimal' configuration used by Bambacht and measure it?   ○ Next, implement and perform token performance measurements:    § Throughput    § Double spending instigator detection.  - Performed a new cryptography measurement that says more about the system than the old.

**Comment by KoningR** [comment]
[Thesis_Eurotoken 05-09-22.pdf](https://github.com/Tribler/tribler/files/9487454/Thesis_Eurotoken.05-09-22.pdf)

**Comment by synctext** [comment]
- Your results are quite strong and in-depth. You explored IPC, java, UDP, etc. thus 
- solid foundations for: "disaster-proof digital offline Euro - first realisation"

**Comment by synctext** [comment]
This thesis is _extremely_ relevant for society. Only open source and transparent technology, not full frontal competition against Amazon. https://www.ecb.europa.eu/paym/intro/news/html/ecb.mipnews220916.en.html
https://worldline.com/en/home/pressroom/press-releases/2022/pr-2022_09_16_01.html
```
Worldline has been selected for the specific use case “peer-to-peer offline payments” of a digital euro, which focuses
on the payment between individuals.
As a leader in payment services, Worldline can rely on its expertise and assets to build a digital wallet supporting
the physical storage of funds that can be transferred without connectivity. The aim of the prototyping exercises is
to test how well the technology behind a digital euro integrates with various use-cases.
Worldline is sharing the common goal of the ECB and its partners and intends to be an active participants of
payment industry evolution, by contributing to strategic and potentially transformative projects such as the
digital euro. Its entire corporate product portfolio can be leveraged to deliver pilots and facilitate a successful
CBDC roll-out
```

**Comment by KoningR** [comment]
### Sprint log

- Changed terms in figures slightly to be more finance-oriented.
- Incorporated feedback of last week.
- Combined existing results of cryptography and UDP throughput to give a more meaningful picture.
- Modified the figures to display the combined results.
- Rewrote problem description to be not-Wessel-oriented.
- Wrote about possible replay attack and possible solutions.
- Decided against leaning too heavily into politics, because I do not feel comfortable with my knowledge on the subject.
- Wrote initial discussion/future work.
- Read the articles posted on this ticket last week; incorporated two sources "to get the reader into a "this is really real" mode of thinking".

**Comment by KoningR** [comment]
[Thesis_Eurotoken_09_26_22.pdf](https://github.com/Tribler/tribler/files/9643853/Thesis_Eurotoken_09_26_22.pdf)

**Comment by synctext** [comment]
- cash is king argument. Cash is dying in Sweden ([leading in EU](https://www.riksbank.se/en-gb/payments--cash/payments-in-sweden/payments-report-2021/1.-trends-on-the-payment-market/pandemic-hastening-development-towards-digital-payments/cash-used-less-and-less-frequently-in-sweden-and-abroad/)) <IMG src=https://user-images.githubusercontent.com/325224/192218284-f7e7f630-d326-4c0f-836a-65053f4ae5f7.png width=300>

**Comment by KoningR** [comment]
https://www.ecb.europa.eu/paym/digital_euro/investigation/profuse/shared/files/dedocs/ecb.dedocs220422_keymilestones.en.pdf

Slide 2

**Comment by synctext** [comment]
[Australian CBDC Pilot for Digital Finance Innovation](https://dfcrc.com.au/wp-content/uploads/2022/09/RBA-DFCRC-CBDC-Research-Project-Australian-CBDC-Pilot-for-Digital-Finance-Innovation-White-Paper.pdf)

**Comment by KoningR** [comment]
### Sprint log
- Decided to not lean on the 'cash is king' argument. Why? Because that is not the reason for a CBDC. It appears that central banks care little about the practical benefits of using cash: its usage has been in decline for  well over a decade, yet a CBDC is only being discussed in the last few years. Then what is the reason? The rise of other payment services and CBDCs.
- Cleaned up thesis document and rewrote (parts of) every section.
- Added reference to Bitcoin paper but did not write about it in related work because it is not really related.
- Read various papers on the developments of e-cash. 


### (Some) Papers Regarding Digital Cash
- Chaum, Blind signatures for untraceable payments, paper introducing digital cash and blind signatures.
- Canard, Anonymity in transferable e-cash, paper introducing strongly unlinkable e-cash and impossibility results. Canard, Divisible e-cash made practical, paper introducing secure divisible e-cash.

**Comment by KoningR** [comment]
[Thesis_Euro_14_10_22.pdf](https://github.com/Tribler/tribler/files/9783451/Thesis_Euro_14_10_22.pdf)

**Comment by synctext** [comment]
You discovered that your [thesis field has 30 years of history - 77 page overview](https://eurocrypt.iacr.org/2018/Slides/Wednesday/InvitedTalk.pdf)
Year | System of cybercurrency
---|---
1983 | [Blind signatures by Chaum](http://www.hit.bme.hu/~buttyan/courses/BMEVIHIM219/2009/Chaum.BlindSigForPayment.1982.PDF) - 6-page basic idea
1989 | digicash http://iwai-k.com/HowToCirculateECash.pdf
1996 | [e-Gold - real gold bars with ownership tokens](https://cs.stanford.edu/people/eroberts/cs201/projects/2010-11/Bitcoins/e-gold.html)
2002 | HashCash http://www.hashcash.org/hashcash.pdf
2005 | decentralized [digital currency by Nick Szabo](https://nakamotoinstitute.org/bit-gold/)
2007 | [bandwidth as a currency](http://news.bbc.co.uk/2/hi/technology/6971904.stm) by Tribler lab
2008 | Bitcoin initial idea https://satoshi.nakamotoinstitute.org/emails/cryptography/1/

**Comment by synctext** [comment]
Path to completion of master thesis:
- [X] All code is complete, offline clean-slate coin 
- [X] Scope is clearly defined as: _Implementing Offline Digital Currency on IPv8_. Nothing else needs to be included, finished or in the way of graduation.
- [ ] Additional measurements, insight in transaction rate, "Section called EVA mystery" {that means you do not bear the reposonsibility of fixing this} , and performance breakdown {IPC, huge cost of crypto}
- [ ] Polish writing and expand from current 6-pages to 8 or 10 pages.
- [ ] DONE, **_really_** DONE {pick date, find committee member, make slides, hold talk}

**Comment by synctext** [comment]
Detailed ToDo list for sections:
- [ ] "Nevertheless, there is demand for this reliability, especially in times of crisis". Essential concluding point is missing that: "CBDC in the form of a digital Euro fix the systemic risk that numerous commercial banks can go bankrupt in a large-scale financial crisis." By design, the digital Euro can't go bankrupt or will be the last to default.

**Comment by KoningR** [comment]
[Thesis_Eurotoken_25_11_22.pdf](https://github.com/Tribler/tribler/files/10094373/Thesis_Eurotoken_25_11_22.pdf)

 - Which related works are related/important 'enough'?
     ○ Currently discussed in Related Work section:
        § First digital cash (Chaum)
        § First transferable and divisible cash
        § First 'fair' digital cash (i.e. judicial powers have the possibility of inspection) 
        § Bitcoin
        § Eurotoken

    ○ Discussed in mentioned presentation (though I do not consider these particularly relevant for my thesis, mostly historically):
        § E-Gold, Liberty Reserve
        § B-Cash, HashCash
        § Zcash, Ethereum, Cardano, Iota
   
    ○ Ideas for additional related work:
        § Digital cash with the strongest notion of privacy.
        § An (algorithmic) stablecoin
 
- Text-related TODO items not done (yet): 
   ○ Trusted cryptographic coprocessor in Related Work.
   ○ 7+ related entries from presentation

**Comment by KoningR** [comment]
- Modified thesis text (see previous post). 
 - Performed a parameter search over window and block sizes of EVA.
       - Larger block sizes could not be used because EVA would crash.
       - EVA failed so often that this couldn't be automated. Perhaps worth mentioning in thesis text?
       - Performed all measurements again with encryption disabled to measure its impact on performance: encryption does very little.

<img src="https://user-images.githubusercontent.com/23524378/205510838-7e6d24d9-cba2-4746-9263-dd5d87b610c5.png" width="400" height="400" />

**Comment by synctext** [comment]
- [ ] Introduction almost finished. "They cannot safeguard a reliability comparable
to that of ECB-backed cash.", please make stronger. Just mention 1989 and 2008 events: Every few decades a severe financial crisis erupts which leads to bankruptcy, multiple bank collapses, and risk of a financial system crash.

**Comment by KoningR** [comment]
### List of completed TODOs (previous iteration and current)
- ""Nevertheless, there is demand for this reliability, especially in times of crisis". Essential concluding point is missing that: "CBDC in the form of a digital Euro fix the systemic risk that numerous commercial banks can go bankrupt in a large-scale financial crisis." By design, the digital Euro can't go bankrupt or will be the last to default."
    - CHECK Wrote: "Euro cash is the only public form of money that is directly backed by the ECB. Digital payments are not; they are backed by private parties such as commercial banks. These parties cannot safeguard reliability comparable to that of ECB-backed cash. Nevertheless, there is demand for reliability, especially in times of crisis \cite{ecb_2021_banknotes}. CBDCs can provide this reliability and safeguard consumers against the effects of large-scale bankruptcy of commercial payment providers."
	
- "{repeating} Stronger intro argument example: The decline of cash is eroding the sovereignty of the Euro. The European Central Bank is responsible for cash. Banks create money by lending. Electronic money vastly outnumbers cash in circulation. Almost all economic transaction are becoming vulnerable for bankruptcy of financial actors. The decline of cash leads to systemic system instability."
    - UNCHECK As discussed previously, the need for a digital Euro is not only due to commercial banks. Moreover, central banks do not intend to replace commercial banks. Therefore, this is not a suited opening argument.
    - Instead, I wrote: "Euro cash is the only public form of money that is directly backed by the ECB . Digital payments are not; they are backed by private parties such as commercial banks. A critical dependence on these parties is eroding the sovereignty of the Euro. They cannot safeguard reliability comparable to that of ECB-backed cash. Nevertheless, there is demand for reliability, especially in times of crisis. In recent history, there have been several financial crises that caused large-scale bankruptcies which consequently impacted consumer's savings (e.g. in 2008). CBDCs can provide reliability and safeguard consumers against the effects of large-scale bankruptcy of commercial payment providers."

- "There have been numerous attempts at solving or mitigating the double spending problem. " suggestion: The double spending problem has never been solved within the offline setting, only within an online context". keep remainder of existing text."
    - CHECK Wrote: "The double spending problem has never been solved in an offline setting, only in an online setting."

- "Related work: "An early prototype solution from 2005 to the double spending problem relied on trusted computing. This experimental work critically relied on a centralised high-quality secure processor, the IBM 4758 PCI Cryptographic Coprocessor. ref""
    - UNCHECK Wrote about different related work.

- "Related work: numerous theoretical proposals have been made for transferable coins, fungible coins, anonymous coins, and coins with conditional privacy. Many of these historical proposals are not relevant directly for our digital Euro, as we assume some institutional trust."
    - CHECK Wrote about different types of e-cash
   
- ""They cannot safeguard a reliability comparable
	to that of ECB-backed cash.", please make stronger. Just mention 1989 and 2008 events: Every few decades a severe financial crisis erupts which leads to bankruptcy, multiple bank collapses, and risk of a financial system crash."
    - CHECK Wrote "In recent history, there have been several financial crises that caused large-scale bankruptcies which consequently impacted consumer's savings (e.g. in 2008)."

- ""Meta, which initiated Diem" Mention Facebook (now named Meta)"
    - CHECK Wrote: "An interested commercial party is for instance Meta (formerly Facebook)"

- ""CBDC to be usable in an offline environment", citation needed.
    - CHECK Added a source.

- "{Add citation} As we both agree: Our [offline-first architecture is resilient against failing central servers](https://github.com/offlinefirst/research/blob/master/links.md). The work presented here has a more robust architecture when compared to current commercially deployed systems. Consumers experience regular failings with Internet banking and payments. A consumer watchdog TV program concluded that: "bank systems are unreliable"[REF](https://www.bnnvara.nl/kassa/discussie-systemen-bij-banken-niet-betrouwbaar)."
    - CHECK Wrote: "The currency can be spent offline and guarantees retroactive fraud detection. It is therefore resilient against temporary failure of central servers, unlike many currently deployed systems."
     - UNCHECK I could not find a proper source about regular failings of internet banking and payments. The provided source is Dutch and 10 years old.

- "{clarity} "Double-spending is defined as the process of making two payments with the same currency or funds in order to deceive the recipient of those funds. ""
    - CHECK Wrote: "The main difficulty with implementing offline digital currency is the double spending problem. Double spending is the action of spending a digital unit of value more than once, illegitimately. In a digital environment this is possible because currency is easily duplicated."
	
- "Two approaches are possible, a balanced-based design and a coin-based design. move all details here: "Eurotoken is a balance-based system where individual units of currency are unnamed""
    - CHECK See Problem Description.

- "Move to the correct subsection: "Up until then, electronic cash had been non-transferable. Nontransferable e-cash can be spent only once, after which it must be redeemed by a trusted authority.""
    - CHECK

- "Subsections are too small, each section has a nice italics topic , remove headings: "B. Transferable and divisible electronic cash""
    - CHECK

- "Bitcoin == 2009"
    - CHECK But 2008*.

- ""We consider the main prior work for this thesis to be Eurotoken [16], another digital currency implemented on IPv8", more concrete like: 'We consider the main prior work for this thesis to be the first deployed Eurotoken prototype by Delft University of Technology [16], this digital currency is also implemented on IPv8. This first prototype is balance-based, instead of our second coin-based prototype."
    - CHECK Wrote: "We consider the main prior work for this thesis to be the first Eurotoken prototype by Delft University of Technology. This digital currency is also implemented on IPv8.". However, moved the comparison between balance- and coin-based to the end of the section to avoid unnecessary context switches.

- "Wessel,Robbert, Wessel storyline. "Eurotoken is a balance-based system where individual units
	of currency are unnamed", avoid 1 context switch."
    - CHECK

**Comment by KoningR** [comment]
[Thesis_Eurotoken_15_12_22.pdf](https://github.com/Tribler/tribler/files/10239992/Thesis_Eurotoken_15_12_22.pdf)

**Comment by KoningR** [comment]
<img src="https://user-images.githubusercontent.com/23524378/207948637-c2fa0bc2-4f4f-442c-8548-ead623ae975b.png" width="400" height="400" />

**Comment by synctext** [comment]
- _interesting discussion_ Add the current societal debate. This is nothing new: see debate on [too big to fail and remove the consumer involvement, make it a normal industry](https://en.wikipedia.org/wiki/Separation_of_investment_and_retail_banking). However, a critique on the current approach is that nothing essential might change. It's still the commercial bank who will operate the digital Euro, just now with the backing of the central bank.

**Comment by KoningR** [comment]
## List of completed TODOs

 - **Solved the 'EVA mystery'. Loss of performance is due to buffer overflows.**
     - The following plots show a steep decline in throughput for large window sizes and a corresponding increase in packet loss. When the buffers are increased, throughput increases and packet loss diminishes.
     - Regarding parameters: More than ~1200 bytes per block is not supported by EVA and Kotlin-IPv8. More than 256 blocks per window or buffer sizes smaller than 100k bytes caused EVA to crash too often to perform proper measurements.
     - <img src="https://user-images.githubusercontent.com/23524378/211308166-4d3530af-822b-4a35-96de-6eca48da0c26.png" width="300" height="300" />
     - <img src="https://user-images.githubusercontent.com/23524378/211308350-9ba705b6-5150-4fba-835d-82e49c74b2de.png" width="300" height="300" />
     - <img src="https://user-images.githubusercontent.com/23524378/211309302-09ae5667-008f-42b3-9a15-c51a0bb522b9.png" width="300" height="300" />

 - Fixed a *race condition* that caused EVA to fail frequently; it is now stable enough to perform N repetitions subsequently without human intervention (restarts etc.).
     - **Can I write about this in my thesis?**
     - The protocol would execute its send function ```sendScheduled()``` both from its internal scheduler as well as when a transfer was finished. It would occur that ```sendScheduled()``` executed from multiple threads simultaneously and thereby start multiple simultaneous transfers. The receiving client would then throw an error, as it can only handle a single transfer at a time. Usually, this error would go unnoticed because the internal scheduler would have a frequency that is low, such that the chances of the problem occurring are low. When doing a performance analysis however, this frequency needs to be increased massively to stabilize results.

- Regarding thesis text ([Thesis_Eurotoken_09_01_22.pdf](https://github.com/Tribler/tribler/files/10373237/Thesis_Eurotoken_09_01_22.pdf))
    - Offline First
        - Mentioned 'Offline First' at the end of the introduction. However, could not find a proper source discussing the paradigm. Opting to remove it if no proper source can be found.
    - Overspending storyline
        - Discussing the transposition of the double spending problem into an overspending problem is not relevant for my thesis because current implementation raises a warning when double spending occurs and does not properly support 'overspending' i.e. spending with future funds.
    - Public support
        - Mentioned public support at the end of the problem description. 
    - (Monetary) authority
        - Changed 'authority' to 'monetary authority' in the introduction, problem description, and related work sections. In the design and evaluation sections we still refer to 'authority' for brevity but we have included a small disclaimer in the opening paragraph.
    - Performance analysis
        - Described that we built a prototype.
    - Price stability
        - Moved to Related Work and mentioned out of scope.
    - Conclusion and Future Work
        - Wrote a new conclusion and discussed future work briefly. 
    - Figure 1
        - Recreated Figure 1 with less technical detail but more clarity.

**Comment by synctext** [comment]
Thesis review remarks:
- [ ] "An additional lesson from the literature is that robust realisations are lacking for numerous theoretical proposals made over the last 39 years." mention actual topic: coins, balance, accounts, etc.

**Comment by synctext** [comment]
- Making solid thesis progress. Good to read!
- Good idea. Like: Evaluated three cryptographic libraries within this thesis. This section shows the vast performance superiority of the Libsodium, compared to others. We believe the singleton programming pattern is responsible for the lower performance within Bouncycastle and I2P (invisible Internet project, home made crypto).

**Comment by KoningR** [comment]
[Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_13_02_23.pdf](https://github.com/Tribler/tribler/files/10721496/Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_13_02_23.pdf)

**Comment by synctext** [comment]
- ```D. Tokens```, more descriptive titles versus token and EVA.
- ```Fig. 7. Throughput (left) and packet loss (right) for various configurations of EVA.``` try making this into a readable figure. For instance, just with 5 lines for each of the block-sizes.

**Comment by InvictusRMC** [comment]
[Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_20_02_23.pdf](https://github.com/Tribler/tribler/files/10814435/Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_20_02_23.pdf)

Hi Robbert,

I think you're well on your way. I've written some of my thoughts down and highlighted some typos. I think your work could benefit from something like a _system & threat model_ section, as it is not clear what kind of network topology is used or what (malicious) actors are capable of. Good luck!

**Comment by KoningR** [comment]
[Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_06_03_23.pdf](https://github.com/Tribler/tribler/files/10896434/Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_06_03_23.pdf)

- Made two PRs for improvements to Kotlin-IPv8.

**Comment by synctext** [comment]
- Kotlin-IPv8 was unmaintained for a year. Instead of Quick or uTP we first need to get a [42 years old TFTP protocol](https://www.rfc-editor.org/rfc/rfc783.html) dubbed EVA operational {with sliding window extension}.
- ```Fig. 3. The benefits of parallelism diminish quickly``` You have 1.2 CPU's ?

**Comment by KoningR** [comment]
[Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_13_03_23.pdf](https://github.com/Tribler/tribler/files/10955302/Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_13_03_23.pdf)

**Comment by synctext** [comment]
- Political traction of CBDC, digital Euro, and helicopter money. EU Commissioner level, [Speech by Commissioner McGuinness at the European Parliamentary Financial Services Forum on a digital euro](https://ec.europa.eu/commission/presscorner/detail/en/SPEECH_23_1547)

**Comment by KoningR** [comment]
[Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_20_03_23.pdf](https://github.com/Tribler/tribler/files/11015490/Performance_Analysis_of_an_Offline_Digital_Euro_Prototype_20_03_23.pdf)

**Comment by synctext** [comment]
- Made flame graph, but not yet included because it duplicates the 'IPv8 dominates' messages.
- Please include a flame graph, it strengthens the message.
-

**Comment by KoningR** [comment]
Very brief minimal content of presentation:

 - Short introduction
     - ECB to Kotlin offline
     - Skip most of problem description & related work
  - Figure 1, token format.
  - Elaborate on token formatting (section 3)
      - Emphasize double spending prevention
  - Elaborate on performance analysis
      - Figure 2 Loss of performance in Kotlin, EVA (figure 7)

**Comment by KoningR** [comment]
[Performance Analysis of an Offline Digital Euro Prototype.pptx](https://github.com/Tribler/tribler/files/11099016/Performance.Analysis.of.an.Offline.Digital.Euro.Prototype.pptx)

**Comment by synctext** [comment]
[Latest ECB wallet report](https://www.ecb.europa.eu/press/pr/date/2023/html/ecb.pr230424_1_annex~93abdb80da.en.pdf)
[Latest report by Dutch minister to parliament](https://www.rijksoverheid.nl/documenten/kamerstukken/2023/05/16/kamerbrief-mogelijke-vormgeving-en-planning-digitale-euro) May 2023

**Comment by milahu** [comment]
how is CBDC related to tribler?

> towards proven technology, ready for large-scale deployment

CBDC is orthogonal to tribler's goal of censorship-resistance.
CBDC is a tool for totalitarian control and censorship, similar to china's social credit score system.
CBDC is NOT a blockchain-based cryptocurrency like bitcoin/monero/zerocoin/..., but just a different version of fiat money.

they are trying to impose CBDC by force, but at least in africa ([nigeria](https://www.nakedcapitalism.com/2023/06/central-bank-of-nigerias-governor-who-tried-to-entrench-a-cashless-economy-has-been-suspended-and-arrested.html)), people are waking up.
